### ISSUE TITLE
Create a ZIM file of WikiHow

### ISSUE THREAD

**Issue by kelson42** [openzim/mwoffliner#10]
This would be really useful

**Comment by kelson42** [comment]
This should be really straight forward one time we have #2 done.

**Comment by PrasannaVenkadesh** [comment]
Thanks for inviting me here as well as creating this issue. I have currently cloned this repository and setting up all the dependencies to make use of this. This is the first time I a using mwoffline tool, so expect more comments from me.

**Comment by PrasannaVenkadesh** [comment]
I have managed to setup dependencies for mwoffliner, but I couldn't get the expected version (>=1.3) installed. I have two machines, Archlinux and Ubuntu 14.04 with both contains packages for xapian 1.2.x. I then downloaded xapian 1.4.x source. Compiled and installed, but couldn't find xapian-config in it. Therefore  I had to compile zimwriterfs without xapian.

My first attempt to hit wikihow.com

    ./mwoffline --mwUrl=https://wikihow.com/ --adminEmail=MY_EMAIL

got 404 and the script failed. This is because mwoffliner expects the /wiki URL pattern, while wikhow doesn't follow this pattern. Any hints to proceed further? Have we faced similar issues earlier?

If not, I need to write a custom script for crawling wikihow.com for html contents.

**Comment by kelson42** [comment]
mwoffliner is there to mirror a recent Mediawiki instance (ie. with visual editor/Parsoid installed). These pieces provide the necessary backend and provide the needed HTML for mwoffliner. Wikihow does not provides this (for now) and that is why we do not have provided a ZIM file until now. So wee need to have a local Parsoid running, able to deal with Wikihow API provided here: http://www.wikihow.com/api.php.

What is Parsoid and how to install it is here https://www.mediawiki.org/wiki/Parsoid#Parsoid_setup. See also #2.

One time the local Parsoid works correctly with Wikihow API, you should have a call like:
` ./mwoffliner.js --mwUrl=http://www.wikihow.com/ --mwWikiPath="" --mwApiPath="/w/api.php" --parsoidUrl=your_local_url --adminEmail=foo@bar.net`

**Comment by PrasannaVenkadesh** [comment]
@kelson42 thanks for the valuable information. Let me try these and get back to you shortly.

**Comment by kelson42** [comment]
@PrasannaVenkadesh Any news on this? We have also other wikis which would benefit of this!

**Comment by PrasannaVenkadesh** [comment]
@kelson42 Sorry I took too long to get back. I have successfully setup parsoid locally and the web-service is accessible at localhost port 8000 therefore I invoked mwoffliner like the following

`./mwoffliner.js --mwUrl=http://www.wikihow.com/ --mwWikiPath="" --mwApiPath="/api.php" --parsoidUrl=http://localhost:8000 --adminEmail=MY_EMAIL_ID_HERE`

I could see continuous network activity in my router and apart from few exceptions on some URLs like `ECONNRESET, ENETUNREACH`, the script doesn't log anything. Probably I should have started it in debug mode.

Two directories are created, `cac` and `out`. Out is empty (the script is still running since 30 minutes) and cac consists of a sub-directory named `wikiHow_en` inside which a  file named `ref` with value `42` in it and another empty directory `m`.

Using wireshark, I inspected the request and response sent and mostly I see a `200 OK` response and since we requested json data, I could see the content-type in response header and the response itself contains values like the following

`{"warnings": {"main": {"*": "Unrecognized parameter: 'rawcontinue'"}}, "limits": {"backlinks":500}, "query":{"backlinks":[]}}`

Should I keep the script running or is something going wrong here?

**Comment by kelson42** [comment]
Nice. Yes, let run it. The "tmp" and "cac" directory should grow in size.

**Comment by kelson42** [comment]
@PrasannaVenkadesh any news on this?

**Comment by yohanboniface** [comment]
We (at Library Without Borders) are also very intersted by having Wikihow bundled for Kiwix. :)
Is any way we can help in such process?

**Comment by kelson42** [comment]
See also #27

**Comment by PrasannaVenkadesh** [comment]
> @PrasannaVenkadesh any news on this?

Yes. The script ran the entire night. I do not know the exact time when it had terminated, but I woke in the morning to see that the script has terminated. I have taken screenshot of the same. I will share it as soon as possible.

The result is all those folders (cac, tmp & out) remain unchanged.

**Comment by PrasannaVenkadesh** [comment]
@kelson42 any means to debug?

**Comment by kelson42** [comment]
@PrasannaVenkadesh It seems to much requests have been sent to the backend (HTTP 429) error, probably the scrapper should go more slower. How do the scrapped page look like?

**Comment by kelson42** [comment]
mwoffliner has a --speed paramter, May you retry with --speed=1 again?

**Comment by PrasannaVenkadesh** [comment]
> mwoffliner has a --speed paramter, May you retry with --speed=1 again?

Yes. I can run it again.

> How do the scrapped page look like?

Where should I look for? I don't find any scrapped page saved in any of those folders.

**Comment by kelson42** [comment]
@PrasannaVenkadesh Then, that means this is "stuck" by getting redirects. Could you share here your parsoid localsettings.js and your mwoffliner call, so I can try by myself?

**Comment by PrasannaVenkadesh** [comment]
> Could you share here your parsoid localsettings.js

I found two locations for this file.
1. `/etc/mediawiki/parsoid/settings.js`
2. `/usr/lib/parsoid/src/localsettings.js`

I hope the [1] is what we need and since github doesn't allow me to attach js files here, I am saving it as txt and attaching it. Please take a look at [settings.txt](https://github.com/kiwix/mwoffliner/files/635839/settings.txt)

> your mwoffliner call,

    sudo ./mwoffliner.js --mwUrl=http://wikihow.com --mwWikiPath="" --mwApiPath="/api.php" --parsoidUrl=http://localhost:8142 --adminEmail=<email_address>

I am running parsoid on port **8142**

**Comment by kelson42** [comment]
@PrasannaVenkadesh Pretty busy with other stuff but not forgetting this. Could you please upload your localsettings.js, this is the one you need to modify to configure your local Parsoid so it can deal with the remote Wikihow API,

**Comment by Cyberman112** [comment]
Yes. This would be very useful. Having an offline internet how-to guide in ZIM format would be very helpful to survival fanatics that think the internet is going to go down. :D But seriously it could really help families that don't have broadband and only have access through satellite.

Have you considered asking to put the final version on downloads.kiwix.org so that we can all take a look?

**Comment by qpkorr** [comment]
Hi! I'm another person interested in this - and tried to follow the instructions, but got some errors:
sudo node ./node_modules/mwoffliner/bin/mwoffliner.script.js --mwUrl=http://www.wikihow.com/ --mwWikiPath="" --mwApiPath="api.php" --parsoidUrl=http://localhost:8142 --adminEmail=MY_EMAIL
Unable to download content [1] http://www.wikihow.com//Mediawiki:offline.css?action=raw (statusCode=404).
Unable to download content [2] http://www.wikihow.com//Mediawiki:offline.css?action=raw (statusCode=404).
Unable to download content [3] http://www.wikihow.com//Mediawiki:offline.css?action=raw (statusCode=404).
Absolutly unable to retrieve async. URL: Unable to download content [3] http://www.wikihow.com//Mediawiki:offline.css?action=raw (statusCode=404).
Unable to download content [1] http://localhost:8142Main_Page?oldid=16421456 (request error: Error: connect ECONNREFUSED 127.0.0.1:80 ).
Unable to download content [1] http://localhost:8142Accept_the_30_Year_Old_Milestone?oldid=21415988 (request error: Error: connect ECONNREFUSED 127.0.0.1:80 ).
Unable to download content [1] http://localhost:8142Access_Gmail_on_Desktop_Email_Software?oldid=20127059 (request error: Error: connect ECONNREFUSED 127.0.0.1:80 ).
Unable to download content [1] http://localhost:8142180_on_a_Skateboard?oldid=21422208 (request error: Error: connect ECONNREFUSED 127.0.0.1:80 ).
Unable to download content [2] http://localhost:8142Main_Page?oldid=16421456 (request error: Error: connect ECONNREFUSED 127.0.0.1:80 ).
.... (more of the same)

Any suggestions @kelson42 ?

**Comment by kelson42** [comment]
@qpkorr We have done meanwhile a bit of progress on this. The main point is that you can not mirror WH without having a local parsoid. This options has been implemented by @ISNIT0 two weeks ago. That said, we have remarked that WH has a few particularities which more or less avoid to create a ZIM file using a std solution. That is why a custom solution, based on mwoffliner, is planned to be developed to finally achieved this.

**Comment by qpkorr** [comment]
Thanks very much @kelson42 - I guess I'll await further news.  I can't offer a lot of time to help - but could perhaps do some testing if desired.

**Comment by ISNIT0** [comment]
👍

**Comment by stefangrotz** [comment]
Hey everyone, I am interested in this as well. I can do some testing too if that helps.

**Comment by phlame64** [comment]
Would love to see this done! Any way I can help?

**Comment by kelson42** [comment]
@phlame64 Yes, you can start the project. WikiHow is a Mediawiki but we have a lot here of tuning to do to do it right. We need to reuse mwoffliner (so nodejs solution) and then do the necessary tuning.

@ISNIT0 Any additional thoughs? I thought you might have a bit of code to bootstrap this?

I will close this ticket as I have created a dedicated code repo (with issue tracker) here:
https://github.com/openzim/wikihow

Please continue the discussion there.

**Comment by kelson42** [comment]
@ISNIT0 @Popolechien  Looks like we could do this now out-of-the-box with `mwoffliner` so i reopen the ticket.
